src/schedule/zephyr_ll.c

```diff
@@ -0,0 +1,296 @@
+	struct list_item *list, *tmp;
+
+// FIXME: locking
+	/*
+	 * We cannot use a mutex because tasks can be added or deleted from
+	 * interrupt context. TODO: use two lists: staging and running. Use the
+	 * lock to access the staging list, only access the running list in this
+	 * function - with no locking. To cancel set the state to CANCEL, here
+	 * check before running with lock held. Then release the lock and run
+	 * the task. If the task is cancelled during running - that's normal.
+	 * After task return take the lock, if cancelled - remove from the list
+	 * and set to FREE.
+	 */
+	spin_lock(&sch->lock);
+	list_for_item_safe(list, tmp, &sch->tasks) {
+		struct comp_dev *sched_comp;
+		enum task_state state;
+		uint32_t flags;
+
+		task = container_of(list, struct task, list);
+
+		if (task->state == SOF_TASK_STATE_CANCEL) {
+			list_item_del(&task->list);
+			task->state = SOF_TASK_STATE_FREE;
+			continue;
+		}
+
+		if (!domain_is_pending(sch->ll_domain, task, &sched_comp))
+			continue;
+
+		task->state = SOF_TASK_STATE_RUNNING;
+		state = task_run(task);
+
+		if (state == SOF_TASK_STATE_COMPLETED)
+			list_item_del(&task->list);
+
+		if (task->state == SOF_TASK_STATE_RUNNING)
+			task->state = state;
+
+		task->start = sch->ll_domain->next_tick;
+	}
+	spin_unlock(&sch->lock);
+}
+
+/* Called once for periodic tasks or multiple times for one-shot tasks */
+static int zephyr_ll_task_schedule(void *data, struct task *task, uint64_t start,
+				   uint64_t period)
+{
+	struct zephyr_ll *sch = data;
+	struct zephyr_ll_pdata *pdata;
+	struct task *task_iter;
+	struct list_item *list;
+	uint64_t delay = period ? period : start;
+	int ret;
+
+	spin_lock(&sch->lock);
+
+// FIXME: locking
+	/* check if task is already scheduled */
+	list_for_item(list, &sch->tasks) {
+		task_iter = container_of(list, struct task, list);
+
+		/* keep original start */
+		if (task_iter == task) {
+			spin_unlock(&sch->lock);
+			return 0;
+		}
+	}
+
+	pdata = task->priv_data;
+	pdata->period = period;
+
+	task->start = domain_time(sch) + delay;
+	if (sch->ll_domain->next_tick != UINT64_MAX &&
+	    sch->ll_domain->next_tick > task->start)
+		task->start = sch->ll_domain->next_tick;
+	task->state = SOF_TASK_STATE_PENDING;
+
+	/*
+	 * Tasks are added into the list from highest to lowest priority. This
+	 * way they can then be run in the same order. Tasks with the same
+	 * priority are served on a first-come-first-serve basis
+	 */
+	list_for_item(list, &sch->tasks) {
+		task_iter = container_of(list, struct task, list);
+		if (task->priority < task_iter->priority) {
+			list_item_append(&task->list, &task_iter->list);
+			break;
+		}
+	}
+
+	/*
+	 * If the task has not been added, it means that it has the lowest
+	 * priority and should be added at the end of the list
+	 */
+	if (list == &sch->tasks)
+		list_item_append(&task->list, &sch->tasks);
+
+	spin_unlock(&sch->lock);
+
+	ret = domain_register(sch->ll_domain, task, &zephyr_ll_run, sch);
```

Comment:
What does this do agian, I guess we also need the unregister too ?

Reply:
oops, indeed we do, thanks for spotting this! In Zephyr case it starts and stops the thread for the first and the last task respectively, and starts and stops the timer for the first and the last task globally over all cores